request.go

```diff
@@ -0,0 +1,88 @@
+package cfdns
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"golang.org/x/exp/maps"
+)
+
+type request[T any] struct {
+	client      *Client
+	method      string
+	path        string
+	queryParams url.Values
+	headers     http.Header
+	body        *T
+}
+
+type response[T any] struct {
+	body    T
+	rawBody []byte
+	code    int
+	headers http.Header
+}
+
+type HTTPError struct {
+	Code    int
+	RawBody []byte
+	Headers http.Header
+}
+
+func (e HTTPError) Error() string {
+	msg := &bytes.Buffer{}
+
+	fmt.Fprintf(msg, "HTTP %d\n", e.Code)
+	headers := maps.Keys(e.Headers)
+	for _, k := range headers {
+		fmt.Fprintf(msg, "%s: %s\n", k, e.Headers.Get(k))
+	}
+	fmt.Fprintln(msg)
+	fmt.Fprintf(msg, "%s", e.RawBody)
+
+	return msg.String()
+}
+
+// IsPermanent returns true if should not try again the same request.
+func (e HTTPError) IsPermanent() bool {
+	if e.Code >= 400 || e.Code < 500 {
```

Comment:
On this case I think it's clearer to use the numbers because I don't care about the specific status codes, I only care if it is a _client error_, and all client errors have the numerical value 4xx. If I apply your suggestion, then you have to know the numerical value of `StatusBadRequest` and `StatusInternalServerError` to understand that I'm looking for 4xx errors.